tests/pyminisketch.py

```diff
@@ -0,0 +1,514 @@
+        # If poly is of the form x+a, its root is a.
+        if len(poly) == 2:
+            return [poly[0]]
+        # Try consecutive randomization factors randv, until one is found that factors poly.
+        while True:
+            # Compute the trace of (randv*x) mod poly. This is a polynomial that maps half of the
+            # domain to 0, and the other half to 1. Which half that is is controlled by randv.
+            # By taking it modulo poly, we only add a multiple of poly. Thus the result has at least
+            # the shared roots of the trace polynomial and poly still, but may have others.
+            trace = poly_tracemod(poly, randv, gf)
+            # Using the set {2^i*a for i=0..fieldsize-1} gives optimally independent randv values
+            # (no more than fieldsize are ever needed).
+            randv = gf.mul2(randv)
+            # Now take the GCD of this trace polynomial with poly. The result is a polynomial
+            # that only has the shared roots of the trace polynomial and poly as roots.
+            gcd = poly_gcd(trace, poly, gf)
+            # If the result has a degree higher than 1, and lower than that of poly, we found a
+            # useful factorization.
+            if len(gcd) != len(poly) and len(gcd) > 1:
+                break
+            # Otherwise, continue with another randv.
+        # Find the actual factors: the monic version of the GCD above, and poly divided by it.
+        factor1 = poly_monic(gcd, gf)
+        factor2, _ = poly_divmod(poly, gcd, gf)
+        # Recurse.
+        return rec_split(factor1, randv) + rec_split(factor2, randv)
+
+    # Invoke the recursive splitting with a random initial factor, and sort the results.
+    return sorted(rec_split(poly, random.randrange(1, 1 << gf.field_size)))
+
+class TestPolyFindRoots(unittest.TestCase):
+    """Test class for poly_find_roots."""
+
+    def field_size_test(self, field_size):
+        """Run tests for given field_size."""
+        gf = GF2Ops(field_size)
+        for test_size in [0, 1, 2, 3, 10]:
+            roots = [random.randrange(1 << field_size) for _ in range(test_size)]
+            roots_set = set(roots)
+            # Construct a polynomial with all elements of roots as roots (with multiplicity).
+            poly = [1]
+            for root in roots:
+                new_poly = [0] + poly
+                for n, c in enumerate(poly):
+                    new_poly[n] ^= gf.mul(c, root)
+                poly = new_poly
+            # Invoke the root finding algorithm.
+            found_roots = poly_find_roots(poly, gf)
+            # The result must match the input, unless any roots were repeated.
+            if len(roots) == len(roots_set):
+                self.assertEqual(found_roots, sorted(roots))
+            else:
+                self.assertEqual(found_roots, [])
+
+    def test(self):
+        """Run tests."""
+        for field_size in range(2, 8):
```

Comment:
Just wondering why this isn't range 2 to 64?

Reply:
Fixed.